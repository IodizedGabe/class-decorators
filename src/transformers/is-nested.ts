import { composePropertyDecorators, IsNestedOptions } from '@hippo-oss/dto-decorators';
import { Type } from 'class-transformer';

import { createBasePropertyDecorators } from './base';

export function IsNested(options: IsNestedOptions): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // convert to nested type
        Type(() => options.type),
    ]);
}
