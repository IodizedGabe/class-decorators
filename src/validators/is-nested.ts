import { composePropertyDecorators, IsNestedOptions } from '@hippo-oss/dto-decorators';
import { ValidateNested } from 'class-validator';

import { createBasePropertyDecorators } from './base';

export function IsNested(options: IsNestedOptions): PropertyDecorator {
    return composePropertyDecorators([
        ...createBasePropertyDecorators(options),

        // validate nested
        ValidateNested({ each: !!options.isArray }),
    ]);
}
