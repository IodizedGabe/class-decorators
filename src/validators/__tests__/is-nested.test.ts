import { validateSync } from 'class-validator';

import { IsNested } from '../is-nested';

describe('IsNested', () => {
    class Child {
    }

    it('allows optional nested to be omitted', () => {
        class Fixture {
            @IsNested({
                optional: true,
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('allows nullable nested to be omitted', () => {
        class Fixture {
            @IsNested({
                nullable: true,
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
    it('does not allow required nested to be omitted', () => {
        class Fixture {
            @IsNested({
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('does not allow non-nested', () => {
        class Fixture {
            @IsNested({
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        obj.property = 'foo' as unknown as Child;
        const errors = validateSync(obj);
        expect(errors).toHaveLength(1);
        expect(errors).toMatchSnapshot();
    });
    it('allows nested', () => {
        class Fixture {
            @IsNested({
                type: Child,
            })
            property!: Child;
        }

        const obj = new Fixture();
        obj.property = new Child();

        const errors = validateSync(obj);
        expect(errors).toHaveLength(0);
    });
});
