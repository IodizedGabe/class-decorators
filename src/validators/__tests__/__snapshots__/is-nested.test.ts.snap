// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IsNested does not allow non-nested 1`] = `
Array [
  ValidationError {
    "children": Array [
      ValidationError {
        "constraints": Object {
          "nestedValidation": "nested property property must be either object or array",
        },
        "property": "property",
        "target": [Function],
        "value": "foo",
      },
    ],
    "property": "property",
    "target": Fixture {
      "property": "foo",
    },
    "value": "foo",
  },
]
`;

exports[`IsNested does not allow required nested to be omitted 1`] = `
Array [
  ValidationError {
    "children": Array [],
    "constraints": Object {
      "isDefined": "property should not be null or undefined",
    },
    "property": "property",
    "target": Fixture {},
    "value": undefined,
  },
]
`;
